Add range deserializer

bhalevy opened the pull request:
Currently, the idl-generated deserialization code, e.g. mutation_partition_view::rows() deserializes and returns a complete utils::chunked_vector<deletable_row_view> . And that could be arbitrarily long.

To consume it gently, we don't need the whole vector in advance, but rather we can consume it one element at a time (and in a nested way for cells in a row in the future).

Use `range_deserializer` to consume range tombstones and rows one item at a time.
    
We may consider in the future also gently iterating over cells
in a row and then dipping into collection cells that might also
contain a large number of items.
    
Fixes #10558

bhalevy commented:
frozen_mutation_test passes on both x86_64 and aarch64
unit(dev) tested on x86 (and in progress on aarch64)

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/342/

avikivity commented on serializer_impl.hh:
```diff
@@ -782,4 +782,138 @@ unknown_variant_type deserialize(Input& in, boost::type<unknown_variant_type>) {
+            : _in(std::move(o._in))
+            , _idx(o._idx)
+            , _item(std::exchange(o._item, std::nullopt))
+        { }
```
Isn't = default (or rather, not defining anything) sufficient? One isn't supposed to use a forward iterator to read the same thing twice, and you aren't preventing it, since the copy constructor and assignment operators are still there.

bhalevy replied:
The problem with the default move constructor is that std::optional move constructor doesn't
make the other optional empty.
https://en.cppreference.com/w/cpp/utility/optional/optional

denesb replied:
Why is that a problem? I don't think we have any requirements w.r.t. dereferencing moved-from iterators.

bhalevy replied:
That's a good point.  I think it can indeed be removed.

avikivity commented on serializer_impl.hh:
```diff
@@ -782,4 +782,138 @@ unknown_variant_type deserialize(Input& in, boost::type<unknown_variant_type>) {
+        bool operator==(const sentinel& s) noexcept {
+            return _idx >= s._idx;
+        }
+
```
const

avikivity commented on serializer_impl.hh:
```diff
@@ -782,4 +782,138 @@ unknown_variant_type deserialize(Input& in, boost::type<unknown_variant_type>) {
+
+        bool operator==(const iterator& it) noexcept {
+            return _idx == it._idx;
+        }
```
const

avikivity commented on serializer_impl.hh:
```diff
@@ -782,4 +782,138 @@ unknown_variant_type deserialize(Input& in, boost::type<unknown_variant_type>) {
+        }
+
+        // Return const& as the input stream cannot be modified
+        const T& operator*() {
```
const

bhalevy replied:
It can't be const since it engages _item.

denesb replied:
You can make `_item` `mutable`. It is only a cache after all. I think the interface's const correctness should reflect the logical state of the object, rather then such implementation details.

bhalevy replied:
return `T&` rather than `const T&` is not really required.
The root problem of defining this method as `const T& operator*() const` is that
it engages `this->_item` and therefore it modifies the iterator.

denesb replied:
Like I said, you can make `this->_item` mutable to work around this problem. Such cache-only members are a typical use-case for mutable. The iterator is mutated logically when it is incremented. Dereferencing doesn't modify its state.

avikivity replied:
input iterators should return T, not const T&.

bhalevy replied:
I agree in theory, but it looks like the standard requires a reference, see:
https://en.cppreference.com/w/cpp/iterator/input_iterator
and https://en.cppreference.com/w/cpp/iterator/indirectly_readable

But anyhow, we'd need to keep the item somewhere for `operator->`, no?

avikivity replied:
I think input iterators typically don't support `operator->`.

denesb replied:
The `LegacyInputIterator` concept requires `operator->()`. See the table of expressions that are required to be valid at https://en.cppreference.com/w/cpp/named_req/InputIterator.

bhalevy replied:
But https://en.cppreference.com/w/cpp/iterator/input_iterator doesn't.
It requires only `*it` (via https://en.cppreference.com/w/cpp/iterator/indirectly_readable)

I'm working on a version returning the item for `operator*` and it is much simpler and elegant.

avikivity commented on serializer_impl.hh:
```diff
@@ -782,4 +782,138 @@ unknown_variant_type deserialize(Input& in, boost::type<unknown_variant_type>) {
+            return *_item;
+        }
+        // Return const* as the input stream cannot be modified
+        const T* operator->() {
```
const

bhalevy replied:
ditto

avikivity commented on serializer_impl.hh:
```diff
@@ -782,4 +782,138 @@ unknown_variant_type deserialize(Input& in, boost::type<unknown_variant_type>) {
+    class iterator {
+        utils::input_stream _in;
+        uint32_t _idx = 0;
+        std::optional<T> _item;
```
Maybe better to keep in range_deserializer. Iterators are presumed cheap to copy.

avikivity replied:
Maybe we don't need it at all. forward_iterator may only be dereferenced once, so it will never be needed to read from this member.

avikivity replied:
s/forward_iterator/input_iterator/

bhalevy replied:
I'm not sure what you meant by not needing it at all.

bhalevy replied:
dereferencing the iterator is supposed to return a T& so it needs to store T somewhere.
If we could just return T we indeed won't need to store it in the iterator.

denesb replied:
https://en.cppreference.com/w/cpp/named_req/InputIterator
`++it` is allowed to invalidate all previously returned references. Any copy of previous `it` values is not even required to be comparable anymore. I think we still need to store the currently de-serialized item locally.

bhalevy replied:
Yeah, I agree. range.begin() can be called more than once so we better keep the iteration state in the iterator rather than in the range.

avikivity replied:
But the T is not part of the iteration state.

bhalevy replied:
std::optional<T> is part of the iteration state (in this version) as it determines if the input_stream should be skipped or not in `operator++`.  If we move std::optional<T> to the range, we'll need to keep an alternative `bool consumed` state in the iterator for this purpose.

Alternatively, we can copy the input stream for the dereference operator so the iterator's input_stream won't be advanced when the item is deserialized, but then `operator++` will always have to skip the item by deserializing the size yet-again, which is wasteful.

avikivity commented on serializer_impl.hh:
```diff
@@ -782,4 +782,138 @@ unknown_variant_type deserialize(Input& in, boost::type<unknown_variant_type>) {
 }
+
+template <typename T>
+class range_deserializer {
```
Please add a comment explaining what this is about.

bhalevy replied:
sure

avikivity commented on serializer_impl.hh:
```diff
@@ -782,4 +782,138 @@ unknown_variant_type deserialize(Input& in, boost::type<unknown_variant_type>) {
+            return sz - sizeof(size_type);
+        }
+
+   public:
```
Please make it a real iterator by adding the standard types.

bhalevy replied:
okay

avikivity commented on idl-compiler.py:
```diff
@@ -1353,6 +1353,15 @@ def param_view_type(t):
+    elif isinstance(t, TemplateType):
+        assert len(t.template_parameters) == 1
+        assert t.name != "boost::variant" and t.name != "std::variant"
+        return param_view_type(t.template_parameters[0])
```
I don't understand why this is called basic_type if it returns something for BasicType || variant.

bhalevy replied:
It's called this way since for `std::vector<T>` it will return `T`.
I'm not sure what would be a better name.  How about `element_type`?

bhalevy replied:
Or `item_type`?

avikivity replied:
element_type should work. But then it should insist on its input being a vector (or other range).

denesb replied:
`value_type`

avikivity commented on idl-compiler.py:
```diff
@@ -1434,6 +1443,18 @@ def add_view(cout, cls):
         """).format(f=DESERIALIZER, **locals()))
 
+        if is_vector(m.type):
+            item_type = basic_type(m.type)
```
What if it's None?

bhalevy replied:
How can it be None if m.type is vector?

avikivity replied:
I misread, somehow I thought the function ended before the is_vector() call, don't know why.

avikivity commented on idl-compiler.py:
```diff
@@ -1434,6 +1443,18 @@ def add_view(cout, cls):
+                  }});
+                }}
+            """).format(f=DESERIALIZER, **locals()))
+
```
Why not replace the existing function?

avikivity replied:
Need unit test for the new function.

bhalevy replied:
> Why not replace the existing function?

It's possible.  I was concerned it could hurt performance, but if the std::optional<T> is moved to the range_deserializer rather than in the iterator that shouldn't be a problem.

> Need unit test for the new function.

`test_writing_and_reading_gently` tests it.
I'll add that to the commitlog.

denesb commented on serializer_impl.hh:
```diff
@@ -782,4 +782,138 @@ unknown_variant_type deserialize(Input& in, boost::type<unknown_variant_type>) {
+        static size_type read_frame_size(utils::input_stream& in) {
+            auto sz = deserialize(in, boost::type<size_type>());
+            if (sz < sizeof(size_type)) {
+                throw std::runtime_error(fmt::format("IDL frame truncated: expected to have at least {} bytes, got {}", sizeof(size_type), sz));
```
Please add context: `range_deserializer::iterator: IDL frame truncated....`. We need to know where this error happened when we see one in the logs.

bhalevy replied:
sure

denesb commented on serializer_impl.hh:
```diff
@@ -782,4 +782,138 @@ unknown_variant_type deserialize(Input& in, boost::type<unknown_variant_type>) {
+    sentinel end() {
+        return sentinel(_size);
+    }
+};
```
Will this work also for `std::unordered_map<K, V>`, making `T=std::pair<K, V>`?

bhalevy replied:
Hmm, not as is,
We may be able to use the same template by defining the view as T=std::pair<K_view, V_view>
Maybe it'd be best to rename the class to `vector_deserializer` until iterative deserialization for maps is implemented.

denesb commented on mutation_partition_view.cc:
```diff
@@ -259,12 +259,12 @@ future<> mutation_partition_view::do_accept_gently(const column_mapping& cm, Vis
     co_await coroutine::maybe_yield();
 
-    for (auto&& rt : mpv.range_tombstones()) {
+    for (auto& rt : mpv.range_tombstones_range()) {
```
I think we should just replace the eager variants with the new lazy variant everywhere. Code wanting to create a copy of the container can do so with the iterator constructor of the respective container.

bhalevy replied:
I'll take a stab at it in the next rev.

bhalevy commented:
In v2:
- renamed to vector_deserializer
- input iterator return value (and simplified)
- added unit test for vector_deserializer
- replaces current vector deserialization

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/416/

denesb commented:
Won't this be able to also work for associative containers with some work, maybe some clever wrapper class passed as T?
We don't have to cover all use-cases right away and of course it can also be renamed later, so this shouldn't block merging.

bhalevy commented:
> Won't this be able to also work for associative containers with some work, maybe some clever wrapper class passed as T? We don't have to cover all use-cases right away and of course it can also be renamed later, so this shouldn't block merging.

Let's consider that as a follow up.  I think that for maps we first need to formalize the element type to be pair<K, V> and then we'll be able to use the vector deserializer to retrieve the pairs and do whatever we want with them.  I'm not sure it needs to be aware, template-wise, that these pairs were serialized from a map since it doesn't produce one back.

denesb commented:
> > Won't this be able to also work for associative containers with some work, maybe some clever wrapper class passed as T? We don't have to cover all use-cases right away and of course it can also be renamed later, so this shouldn't block merging.
> 
> Let's consider that as a follow up. I think that for maps we first need to formalize the element type to be pair<K, V> and then we'll be able to use the vector deserializer to retrieve the pairs and do whatever we want with them. I'm not sure it needs to be aware, template-wise, that these pairs were serialized from a map since it doesn't produce one back.

Yes, follow-up is fine.
